### ISSUE TITLE
Discuss/Question: Is it possible to do windowing with the carousel

### ISSUE THREAD

**Issue by arihantverma** [davidcetinkaya/embla-carousel#120]
Hello!

Is it possible to do windowing, if a carousel has lot of items? Say an image gallery which has a lot of (>300) images

**Comment by davidcetinkaya** [comment]
Hello Arihant (@arihantverma),

Thank you for your question. I'm not quite understanding the feature request. Would you mind sharing an example of this? Like a link to a demo page or similar.

Kindly,
David

**Comment by arihantverma** [comment]
Hey David!

Yes sure!

There are instances on one of our pages in our product, where there are a LOT of DOM nodes on the page. We've tried to mitigate the performance overhead with [react-window](https://react-window.now.sh/). React window shows only those dom nodes that are currently in the browser/parent container viewport and keep hiding the ones that go outside of the container, showing new ones as the user scrolls.

On that page itself, we also have a carousel where a LOT of user generated ( hotel ) images are shown. They range from a couple of hundreds to more in some cases. We were exploring ways in which we could do _windowing_ with embla carousel.

I had been trying to do it with `react-window` itself, but have not yet been able to successfully do it because of the extra html elements and styles that `react-window` injects in order to function properly.

[Here's](https://react-window.now.sh/#/examples/list/fixed-size) an example of horizontal 'windowing' from `react-window` docs. I want to do something similar but with embla.

Does this make sense? If not I'll try to me more clear.

Regards,
Arihant

**Comment by davidcetinkaya** [comment]
Hello again Arihant (@arihantverma),

Thanks for the clarification 👍.

I think what you're asking for is quite achievable with Embla, by utilizing its [API](https://davidcetinkaya.github.io/embla-carousel/api). But I'm not sure it makes sense to add this as a feature to the core, because it will add to the bundle weight and burden users that don't need this feature.

Here's what I would try in order to achieve this:

- Setup a carousel with `loop: true` and a couple of slides, let's say 2 for the sake of this example.
- We replace the content within the slide nodes as the user scrolls and loops the carousel, but only for the one slide that's currently out of view. This can be achieved with the method `embla.slidesInView()`.
- The `embla.scrollProgress()` method lets us track the current scroll progress, so we can use it to track how many times the carousel has looped with a counter variable. Let's say it has looped 1 time and currently slide 1 is visible. In that case we would show the content of the virtual third slide, and so on...

This would be my **To Do** in code:
```ts
let lastScrollProgress = embla.scrollProgress()
let loopCounter = 0

embla.on('scroll', () => {
  // Get the current scroll progress
  // Compare with the last scroll progress to determine scroll direction.
  // If the scroll progress starts over we know that the carousel has looped.
  // In that case, increment or decrement the loop counter variable.
  // Replace the slide content for slides that are out of view.
})
```

Right now, I have tons to do so I don't know when I'll be able to create a CodeSandbox demonstrating this.

Best,
David

**Comment by arihantverma** [comment]
Hey @davidcetinkaya 

Thank you for the explanation! This really helps. I get the point that adding it as a feature doesn't make sense, because that's how embla carousel is made - giving the bare bones api to enable developers to make features on top of it.

No worries that you are occupied, I'll make a codesandbox for this, post it here, close the issue.

Thanks again!

Best
Arihant

**Comment by arihantverma** [comment]
Hey David @davidcetinkaya 

I tried, understood what needs to be done, but haven't been able to wrap my head around how to do it 😔. If you could find time, given that you were busy before, would appreciate your help here, if you could.

Best,
Arihant

**Comment by davidcetinkaya** [comment]
Hi Arihant (@arihantverma),

I understand. Thanks for letting me know. I'll see when I have time to work on this, and will let you know when I have something.

Best,
David

**Comment by cliffordfajardo** [comment]
@arihantverma - I was also thinking about the multiple items per slide.

I have a code sandbox demo here:
- https://codesandbox.io/s/eager-snowflake-7zrpl?file=/src/App.tsx

One way I was able to achieve this to create a utility method to split an array of flat items into sub arrays.

```ts
/*
 * const items = [item1, item2, item3, item4, item5, item6]
 * let slides = createMultiItemSlides(items);
 * slides === [[item1, item2], [item3, item4], [item5, item6]]
 */
export const createMultiItemSlides = <T,>(list: Array<T>, itemsPerSlide: number) => {
  return list.map((item, index, list) => {
    const page = index % itemsPerSlide === 0 ? list.slice(index, index + itemsPerSlide) : null;
    return page;
  })
  .filter((array) => array !== null)
}

/*
 * @description
 * Given a list of items, restructure the original flat list of items
 * into sub arrays, where each sub array will contain `itemsPerSlide`
 * items in them.
 * 
 * The reason why we need this utility is to create a carousel with multiple
 * items per slide.
 * 
 * The problem:
 * Passing a 1 dimensional list like this one: [item1, item2, item3, item4, item5, item6]
 * to our carousel won't work if our goal is to have multiple items per slide.
 * 
 * Most carousel API's  expect you to define the width of each item. Given my flat list of items above,
 *  if I said, make each slide item 33% (3 per page), what will happen is
 * the slide items will have huge gaps between them so they meet the 33% width rule.
 * 
 * Potential solution:
 * Say I want to show 2 items per slide without the huge gaps in between the items and I have an array of 6 items
 * const list = [item1, item2, item3, item4, item5, item6]
 * 
 * I need transform the array to this: [[item1, item2], [item3, item4], [item5, item6]].
 * Each item in the array now represents a slide.
 * 
 * Each slide itself will be 100%. So when I hit 'next' in my carousel, it will shift the entire
 * slide to the next set of items. 
 */
```

**Comment by davidjerleke** [comment]
@kaufmann42,

Thank you for your question. Embla has a [plugin system](https://www.embla-carousel.com/get-started/module/#adding-plugins) so why don't you create a virtualising/windowing plugin? With the Embla API you already have the information about what slides are in view or not:

- [`embla.slidesInView()`](https://www.embla-carousel.com/api/methods/#slidesinview)
- [`embla.slidesNotInView()`](https://www.embla-carousel.com/api/methods/#slidesnotinview)

You can use check out the [embla-carousel-class-names](https://github.com/davidjerleke/embla-carousel/blob/master/packages/embla-carousel-class-names/src/components/index.ts) plugin to see how to structure an Embla plugin.

Best,
David

**Comment by zhondori** [comment]
Helo Everyone.

I made a virtualized carousel using Embla Carousel React.

[Here is the code](https://stackblitz.com/edit/embla-react-virtualization?file=src%2FApp.tsx)

**Comment by IgnisDa** [comment]
@zhondori Would it be possible for your to publish this as a package? Totally understandable if not. Thanks for the code snippet!

**Comment by zhondori** [comment]
> @zhondori Would it be possible for your to publish this as a package? Totally understandable if not. Thanks for the code snippet!

I will try

**Comment by IgnisDa** [comment]
Thanks! That would be awesome! I would be really grateful if you could do that and inform me with the name of the npm package.

**Comment by sahandsn** [comment]
really looking forward for it

**Comment by bpat86** [comment]
Hi @davidjerleke  

Reviving this thread to get your thoughts.

I’ve been watching for a bit, hoping someone else might tackle this, but I decided to take a shot at it myself.

I’ve run into a number of pain points. My first attempt used `slidesInView` and `slidesNotInView` as you outlined in a previous comment, and while it _sort of_ worked, I still ran into the virtual slide sequence issue. Dragging caused some odd behavior as well, swipes or flicks would sometimes jump to unexpected slide indexes, likely due to how the mutation observer is intended to work. Typically it would be off by one.

In my demo, `slidesToScroll` is hardcoded to `3`. In a more thorough implementation, it would be a dynamic value set in state and responsive across breakpoints. If you drag the carousel just enough to reveal the offscreen slide (without triggering a page change), you’ll notice the offscreen slides are of the same set as the current slide group – until the `select` event fires again. It'd be great to have at least one slide from the previous and next pages as a buffer.

I’d like the adjacent slides on either side to appear in proper sequence, but I’m not sure how to accomplish that using Embla’s API. Everything I’ve tried feels hacky. If you have any suggestions, I’d love to hear them as I’ve been stuck on this for a few days. Some ideas I have are to either write some logic to get the next page and swap the first and last slide array items with the correct next slides or append more slides following the `settle` event though I'm not sure how that would affect the existing virtual index logic.

Also, re-implementing the swipe/drag page-change and dot click page-change behavior has been a bit of a headache. Is there a simpler way to preserve the built-in behavior while using a virtual slide implementation, without having to [poorly] recreate it all inside of `watchDrag` or useEffect?

I’ve pulled together the most relevant pieces and cobbled this demo together:

https://codesandbox.io/p/devbox/embla-carousel-slides-per-view-react-forked-xgswdm?workspaceId=ws_AMzbprW7utiAretunhczhR

**Comment by davidmedero** [comment]
Hi folks, first-time commenter here. I’ve been working on slider virtualization in React Motion Gallery and put together a focused handoff in case it’s useful for Embla or a future Embla plugin:

https://github.com/davidmedero/react-motion-gallery/tree/main/packages/react-motion-gallery/docs/embla-slider-virtualization-handoff

The goal is not to ask maintainers to read through our whole React slider. I separated the portable fixed-stride virtual-track algorithm, its tests, and the small set of integration points that seemed relevant to an Embla-style plugin or possible core hook.

The short version: this is fixed-size slider virtualization only for v1, with overscan, threshold fallback, loop index normalization, and scroll-offset-based window updates. Variable-width slides, center alignment, vertical axis, and auto-height are intentionally left out to keep the idea small.

If this direction is interesting, I’d be happy to answer questions, trim the handoff further, or adjust it into whatever format would make review easiest for this repo.
